File: CPMCore/Controllers/SharedController.cs
using Microsoft.AspNetCore.Mvc;
using BOCore;
using FacadeCore;
using System.Linq;


namespace CPMCore.Controllers
{
    public class SharedController : Controller
    {
        private readonly IPostalcodeService _postalcodeService;
        private readonly ICountryService _countryService;
        private readonly IUnitService _unitService;

        public SharedController(IPostalcodeService postalcodeService, ICountryService countryService, IUnitService unitService)
        {
            _postalcodeService = postalcodeService;
            _countryService = countryService;
            _unitService = unitService;
        }

        public IActionResult Index()
        {
            return View();
        }
        [HttpPost]
        public async Task<JsonResult> GetPostcodesByCountry(string term, int CountryId)
        {
            var presponse = await _postalcodeService.GetPostalcodeByCountryAndSearchstring(CountryId, term);

            var postalcodeList = presponse.Success
                ? presponse.Values.Select(x => new Select2DTO
                {
                    id = x.PostcodeId.GetValueOrDefault(),
                    text = $"{x.Postcode} - {x.Gemeente}"
                })
                  .ToList()
                : new List<Select2DTO>();

            return Json(postalcodeList);
        }
        [HttpGet]
        public JsonResult GetPostcodeById(int id)
        {
            var presponse = _postalcodeService.GetPostalcodeById(id);
            if(!presponse.Success)
                return Json(null);

            return Json(new
            {
                Value = presponse.Value.PostcodeId,
                Text = presponse.Value.Postcode + " - " + presponse.Value.Gemeente
            });
        }

        [HttpPost]
        public string GetCountryIsoCode(int countryid)
        {
            var presponse = _countryService.GetCountryById(countryid);
            CountryBO iPostcode = new CountryBO();
            if ((presponse.Success))
                iPostcode = presponse.Values.FirstOrDefault();
            return iPostcode.ISOCode;
        }
        [HttpPost]
        public JsonResult GetAvailableUnitsByProjectId(int id)
        {
            var presponse = _unitService.GetAvailableUnitsByProjectId(id);
            List<IdNameBO> iList = new List<IdNameBO>();
            if ((presponse.Success))
                iList = presponse.Values;
            return Json(iList);
        }

        public class Select2DTO
        {
            // as select2 is formed like id and text so we used DTO
            public int id
            {
                get
                {
                    return m_id;
                }
                set
                {
                    m_id = value;
                }
            }
            private int m_id;
            public string text
            {
                get
                {
                    return m_text;
                }
                set
                {
                    m_text = value;
                }
            }
            private string m_text;
        }
    }
}

File: CPM/Controllers/SharedController.vb
Imports System.Web.Mvc
Imports BO

Namespace Controllers
    Public Class SharedController
        Inherits Controller

        ' GET: Shared
        Function Index() As ActionResult
            Return View()
        End Function
        <HttpPost>
        Public Function GetPostcodesByCountry(term As String, CountryId As Integer) As JsonResult
            Dim pservice = ServiceFactory.GetPostalcodeService()
            Dim presponse = pservice.GetPostalcodeByCountryAndSearchstring(CountryId, term)
            Dim iList As New List(Of PostalCodeBO)
            If (presponse.Success) Then iList = presponse.Values
            Dim PostalcodeList As New List(Of Select2DTO)()
            Dim singlePostalcode As Select2DTO
            For Each selectedPostalcode As PostalCodeBO In iList
                singlePostalcode = New Select2DTO()
                singlePostalcode.id = selectedPostalcode.PostcodeId
                singlePostalcode.text = selectedPostalcode.Postcode & " - " & selectedPostalcode.Gemeente
                PostalcodeList.Add(singlePostalcode)
            Next

            Return Json(PostalcodeList, JsonRequestBehavior.AllowGet)
        End Function
        <HttpPost>
        Function GetCountryIsoCode(countryid As Integer) As String
            Dim pservice = ServiceFactory.GetCountryService
            Dim presponse = pservice.GetCountryById(countryid)
            Dim iPostcode As New CountryBO
            If (presponse.Success) Then iPostcode = presponse.Values.FirstOrDefault
            Return iPostcode.ISOCode
        End Function
        <HttpPost>
        Public Function GetAvailableUnitsByProjectId(id As Integer) As JsonResult
            Dim pservice = ServiceFactory.GetUnitService()
            Dim presponse = pservice.GetAvailableUnitsByProjectId(id)
            Dim iList As New List(Of IdNameBO)
            If (presponse.Success) Then iList = presponse.Values
            Return Json(iList, JsonRequestBehavior.AllowGet)
        End Function

        Public Class Select2DTO
            ' as select2 is formed like id and text so we used DTO
            Public Property id() As Integer
                Get
                    Return m_id
                End Get
                Set(value As Integer)
                    m_id = value
                End Set
            End Property
            Private m_id As Integer
            Public Property text() As String
                Get
                    Return m_text
                End Get
                Set(value As String)
                    m_text = value
                End Set
            End Property
            Private m_text As String
        End Class

    End Class
End Namespace